overlord/servicestate/service_control.go

```diff
@@ -0,0 +1,145 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package servicestate
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap"
+	"github.com/snapcore/snapd/wrappers"
+
+	tomb "gopkg.in/tomb.v2"
+)
+
+// ServiceCommand encapsulates a single service-related action (such as starting,
+// stopping or restarting) run against services of a given snap. The action is
+// run for services listed in names attribute, or for all services of the snap
+// if names is empty.
+type ServiceCommand struct {
+	SnapName       string                 `json:"snap-name"`
+	Action         string                 `json:"action"`
+	ActionModifier string                 `json:"action-modifier,omitempty"`
+	Services       []string               `json:"names,omitempty"`
+	StopReason     snap.ServiceStopReason `json:"stop-reason,omitempty"`
+}
+
+func (m *ServiceManager) doServiceControl(t *state.Task, _ *tomb.Tomb) error {
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	perfTimings := state.TimingsForTask(t)
+	defer perfTimings.Save(st)
+
+	var sc ServiceCommand
+	err := t.Get("service-command", &sc)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot get service-command: %v", err)
+	}
+
+	var snapst snapstate.SnapState
+	if err := snapstate.Get(st, sc.SnapName, &snapst); err != nil {
+		return err
+	}
+	info, err := snapst.CurrentInfo()
+	if err != nil {
+		return err
+	}
+
+	svcs := info.Services()
+	if len(svcs) == 0 {
+		return nil
+	}
+
+	var services []*snap.AppInfo
+	if len(sc.Services) == 0 {
+		// no services specified, take all services of the snap
+		services = info.Services()
+	} else {
+		for _, svc := range sc.Services {
+			app := info.Apps[svc]
+			if app == nil {
+				return fmt.Errorf("no such service: %s", svc)
+			}
+			if !app.IsService() {
+				return fmt.Errorf("%s is not a service", svc)
+			}
+			services = append(services, app)
+		}
+	}
+
+	meter := snapstate.NewTaskProgressAdapterLocked(t)
+
+	switch sc.Action {
+	case "stop":
+		flags := &wrappers.StopServicesFlags{
+			Disable: (sc.ActionModifier == "disable"),
+		}
+		reason := sc.StopReason
+		if reason == "" {
+			reason = snap.StopReasonUserAction
+		}
+		if err := wrappers.StopServices(services, flags, reason, meter, perfTimings); err != nil {
+			return err
+		}
+		// update the list of disabled services; this affects operations such as snap refresh where
+		// snap-control is invoked from hooks (via snapctl), a service is stopped by the hook and should
+		// not then get started by start-snap-services.
+		snapst.LastActiveDisabledServices = append(snapst.LastActiveDisabledServices, sc.Services...)
```

Comment:
bit confused, shouldn't we do this only for Disable: true?  if that's not the case then maybe the Field is a bit misnamed?

Reply:
also afair that value is set only during a refresh otherwise the state is with systemd.  @anonymouse64 can confirm(?)

Reply:
well when I wrote the code LastActiveDisabledServices was set for refresh, revert, and disable. I need to read the rest of the code here to fully respond to your point, I'm not sure if that's still the case that we need to do this, but if all the spread tests for reverting, disabling and refreshing snaps with disabled services are happy then the logic should be correct.

Reply:
yes, with refresh I really meant anything doing unlink/link. The tests are very important but that field has always been confusing if needed to start with, I worry also about breaking invariants in confusing ways, like having it set when it shouldn't, or not meaning what it meant originally.

Reply:
anyway this is not connected to anything yet, so maybe it shouldn't deal with that field quite yet

Reply:
mmh, thinking further, by definition of that field usually if we can talk to systemd about a service that field should not be in play at all, it's needed/was introduced to remember about services that we (temporarely) removed from under systemd control.  If we need to remember some state here it should be about something else?...

Reply:
These are fair questions and concerns. The idea of updating this here is to support cases where hooks manipulate some services during refresh and that should override whatever was remembered for the given service when the operation started.

Thinking about it now, it definitely doesn't need to be introduced in this PR, and maybe I'm overcomplicating this (I need to reconsider the flow). I'll remove it for now and re-introduce in followup PRs where it's actually used.

Reply:
thanks.  As I said:

* I hope either we don't need it, because state is queryable on systemd at this point (we can find out if something is enabled/disabled there)
* at the point we can talk to systemd that field should be *empty*, if it isn't the case we broke already some invariant and we should look into that
* if we really need to keep some state on the side of what is kept by systemd we probably should introduce something else

Reply:
As discussed, this will be address in the followup.